pallets/capacity/src/lib.rs

```diff
@@ -634,22 +953,297 @@ impl<T: Config> Pallet<T> {
+		let current_era_info = CurrentEraInfo::<T>::get();
+		let era_length: BlockNumberFor<T> = T::EraLength::get().into();
+
+		let era_diff = if current_era_info.era_index.eq(&era) {
+			1u32.into()
+		} else {
+			era.saturating_sub(current_era_info.era_index).saturating_add(1u32.into())
+		};
+		current_era_info.started_at + era_length.mul(era_diff.into()) - 1u32.into()
+	}
+
+	// Figure out the history chunk that a given era is in and pull out the total stake for that era.
+	pub(crate) fn get_total_stake_for_past_era(
+		reward_era: RewardEra,
+		current_era: RewardEra,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		// Make sure that the past era is not too old
+		let era_range = current_era.saturating_sub(reward_era);
+		ensure!(
+			current_era.gt(&reward_era)
+				&& era_range.le(&T::ProviderBoostHistoryLimit::get().into()),
+			Error::<T>::EraOutOfRange
+		);
+
+		let chunk_idx: u32 = Self::get_chunk_index_for_era(reward_era);
+		let reward_pool_chunk = ProviderBoostRewardPools::<T>::get(chunk_idx).unwrap_or_default(); // 1r
+		let total_for_era =
+			reward_pool_chunk.total_for_era(&reward_era).ok_or(Error::<T>::EraOutOfRange)?;
+		Ok(*total_for_era)
+	}
+
+	/// Get the index of the chunk for a given era, history limit, and chunk length
+	/// Example with history limit of 6 and chunk length 3:
+	/// - Arrange the chunks such that we overwrite a complete chunk only when it is not needed
+	/// - The cycle is thus era modulo (history limit + chunk length)
+	/// - `[0,1,2],[3,4,5],[6,7,8],[]`
+	/// Note Chunks stored = (History Length / Chunk size) + 1
+	/// - The second step is which chunk to add to:
+	/// - Divide the cycle by the chunk length and take the floor
+	/// - Floor(5 / 3) = 1
+	/// Chunk Index = Floor((era % (History Length + chunk size)) / chunk size)
+	pub(crate) fn get_chunk_index_for_era(era: RewardEra) -> u32 {
+		let history_limit: u32 = T::ProviderBoostHistoryLimit::get();
+		let chunk_len = T::RewardPoolChunkLength::get();
+		// Remove one because eras are 1 indexed
+		let era_u32: u32 = era.saturating_sub(One::one()).into();
+
+		// Add one chunk so that we always have the full history limit in our chunks
+		let cycle: u32 = era_u32 % history_limit.saturating_add(chunk_len);
+		cycle.saturating_div(chunk_len)
+	}
+
+	// This is where the reward pool gets updated.
+	pub(crate) fn update_provider_boost_reward_pool(era: RewardEra, boost_total: BalanceOf<T>) {
+		// Current era is this era
+		let chunk_idx: u32 = Self::get_chunk_index_for_era(era);
+		let mut new_chunk =
+			ProviderBoostRewardPools::<T>::get(chunk_idx).unwrap_or(RewardPoolHistoryChunk::new()); // 1r
+
+		// If it is full we are resetting.
+		// This assumes that the chunk length is a divisor of the history limit
+		if new_chunk.is_full() {
+			new_chunk = RewardPoolHistoryChunk::new();
+		};
+
+		if new_chunk.try_insert(era, boost_total).is_err() {
+			// Handle the error case that should never happen
```

Comment:
I agree that this should never happen, but adding a log message here could be helpful:
log::error!("Failed to insert into the provider boost reward pool for era: {}", era);

Reply:
I wonder if tracing::info is better than log::info for distributed tracing , if at all these places are ever hit